src/auth/mod.rs

```diff
@@ -0,0 +1,171 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types for parsing and authenticating HAWK headers.
+//! Matches the [Python logic](https://github.com/mozilla-services/tokenlib).
+//! We may want to extract this to its own repo/crate in due course.
+
+#[cfg(test)]
+mod test;
+
+use std::{
+    error::Error,
+    fmt::{self, Display, Formatter},
+};
+
+use actix_web::{error::ResponseError, http::header::ToStrError, FromRequest, HttpRequest};
+use base64::{self, DecodeError};
+use chrono::offset::Utc;
+use hawk::{Error as HawkError, Header as HawkHeader, Key, RequestBuilder};
+use hkdf::{Hkdf, InvalidLength};
+use hmac::{
+    crypto_mac::{InvalidKeyLength, MacError},
+    Hmac, Mac,
+};
+use ring;
+use serde_json::{self, Error as JsonError};
+use sha2::Sha256;
+use time::Duration;
+
+use settings::Settings;
+
+static YEAR: u64 = 60 * 60 * 24 * 365;
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct HawkPayload {
+    expires: f64,
+    node: String,
+    salt: String,
+    uid: u64,
+}
+
+impl HawkPayload {
+    fn new(
+        header: &str,
+        method: &str,
+        path: &str,
+        host: &str,
+        port: u16,
+        settings: &Settings,
+        expiry: u64,
+    ) -> AuthResult<HawkPayload> {
+        if &header[0..5] != "Hawk " {
+            return Err(AuthError);
+        }
+
+        let header: HawkHeader = header[5..].parse()?;
+        let id = header.id.as_ref().ok_or(AuthError)?;
+
+        let payload = HawkPayload::extract_and_validate(id, &settings.master_secret, expiry)?;
+
+        let token_secret = hkdf_expand_32(
+            format!("services.mozilla.com/tokenlib/v1/derive/{}", id).as_bytes(),
+            Some(payload.salt.as_bytes()),
+            &settings.master_secret,
+        )?;
+        let token_secret = base64::encode_config(&token_secret, base64::URL_SAFE);
+
+        let request = RequestBuilder::new(method, host, port, path).request();
+        if request.validate_header(
+            &header,
+            &Key::new(token_secret.as_bytes(), &ring::digest::SHA256),
+            Duration::weeks(52),
```

Comment:
I think we should add a comment to explain what this `weeks(52)` is doing here, because is could easily raise eyebrows.